### PR TITLE
fix --watch on multiple requests

### PR DESCRIPTION
**Pull request by juanvallejo**
**Release note**:
```release-note
NONE
```

`kubectl get <resource> --watch` only supports watching a single resource kind at a time.
This check fails if more than one resource `Info` is returned.

When dealing with large quantities of a single resource kind, or an amount that exceeds the value of `--chunk-size`, more than one request is made to the server causing a resource `Info` to be created for each of the requests, ultimately causing the above check to fail even though we are dealing with the same type of resource.

This patch modifies that check to take into account the GVKs of all infos returned, and only fail if at least one differs.

cc @deads2k

### CODE REVIEW

**Comment by deads2k** [comment]
You'll need a test for this.  I'd recommend configmaps

**Comment by deads2k** [comment]
@smarterclayton this feature is your pet.  What did you expect to happen here?

**Comment by smarterclayton** [comment]
I just refactored this last.  Watch on multiple resources seems weird and we don’t have to support it yet.  I’m confused why you are getting multiple infos for a single chunk (or said another way, what does chunk size have to do with this)?

**Comment by juanvallejo** [comment]
@smarterclayton 
> I’m confused why you are getting multiple infos for a single chunk (or said another way, what does chunk size have to do with this)?

When the chunk size is smaller than the overall size of a list of resources, an `Info` is created for each chunk. This is currently a problem if trying to --watch that list

**Comment by juanvallejo** [comment]
@deads2k added a test, ptal

**Comment by juanvallejo** [comment]
@mfojtik @deads2k wondering if this patch looks okay

**Comment by deads2k** [comment]
> @mfojtik @deads2k wondering if this patch looks okay

I'm not very familiar with how the chunking is handled client-side.  I'm surprised it broke.  At the very least, you'll need a test.

File: hack/make-rules/test-cmd-util.sh [hunk 1]
```diff
@@ -1430,6 +1430,16 @@ run_kubectl_get_tests() {
...
   kube::test::if_has_string "${output_message}" "/clusterroles?limit=500 200 OK"
 
+  ### Test kubectl get chunk size does not result in a --watch error when resource list is served in multiple chunks
+  # Pre-condition: no ConfigMaps exist
+  kube::test::get_object_assert configmap "{{range.items}}{{$id_field}}:{{end}}" ''
+  # Post-condition: Create three configmaps and ensure that we can --watch them with a --chunk-size of 1
+  kubectl create cm one "${kube_flags[@]}"
+  kubectl create cm two "${kube_flags[@]}"
+  kubectl create cm three "${kube_flags[@]}"
+  output_message=$(kubectl get configmap --chunk-size=1 --watch --request-timeout=1s 2>&1 "${kube_flags[@]}")
```
Comment (deads2k) [inline hack/make-rules/test-cmd-util.sh]: is it actually related to watch or is related to the initial list?

File: pkg/kubectl/cmd/resource/get.go [hunk 2]
```diff
@@ -463,8 +463,25 @@ func (options *GetOptions) watch(f cmdutil.Factory, cmd *cobra.Command, args []s
 	if err != nil {
 		return err
 	}
-	if len(infos) != 1 {
-		return i18n.Errorf("watch is only supported on individual resources and resource collections - %d resources were found", len(infos))
+	if len(infos) > 1 {
+		gvk := infos[0].Mapping.GroupVersionKind
+		infoCount := 1
```
Comment (deads2k) [inline pkg/kubectl/cmd/resource/get.go]: bad name. `uniqueGVKs`?

File: hack/make-rules/test-cmd-util.sh (hunk 1 above)
Comment (deads2k) [inline hack/make-rules/test-cmd-util.sh]: Add a `--watch-only` test

**Comment by deads2k** [comment]
test makes it clear.  Nit on name

/approve
/assign @soltysh

File: hack/make-rules/test-cmd-util.sh (hunk 1 above)
Comment (juanvallejo) [reply hack/make-rules/test-cmd-util.sh]: related to a check specific to "watch". The check made the assumption that multiple infos equals multiple resource types

**Comment by juanvallejo** [comment]
@deads2k thanks for the feedback, comments addressed

cc @mfojtik

**Pull request merged**